File: src/main/java/com/kishlaly/ta/analyze/tasks/blocks/two/Short_ScreenTwo_EMA_LastBarTooLow.java
package com.kishlaly.ta.analyze.tasks.blocks.two;

import com.kishlaly.ta.model.BlockResultJava;
import com.kishlaly.ta.model.QuoteJava;
import com.kishlaly.ta.model.SymbolDataJava;
import com.kishlaly.ta.model.indicators.EMAJava;
import com.kishlaly.ta.model.indicators.IndicatorJava;
import com.kishlaly.ta.utils.CollectionUtilsJava;
import com.kishlaly.ta.utils.LogJava;

import java.util.List;

import static com.kishlaly.ta.model.BlockResultCodeJava.LAST_BAR_BELOW_SCREEN_2;
import static com.kishlaly.ta.model.BlockResultCodeJava.OK;

/**
 * to filter the situation when the third and second cross EMA13, and the last one is entirely lower (that is, the moment is already lost)
 */
public class Short_ScreenTwo_EMA_LastBarTooLow implements ScreenTwoBlock {
    @Override
    public BlockResultJava check(SymbolDataJava screen) {
        List<EMAJava> screen_2_EMA13 = (List<EMAJava>) screen.indicators.get(IndicatorJava.EMA13);
        QuoteJava quote3 = CollectionUtilsJava.getFromEnd(screen.quotes, 3);
        QuoteJava quote2 = CollectionUtilsJava.getFromEnd(screen.quotes, 2);
        QuoteJava quote1 = CollectionUtilsJava.getFromEnd(screen.quotes, 1);

        boolean thirdCrossesEMA13 = quote3.getLow() < CollectionUtilsJava.getFromEnd(screen_2_EMA13, 3).getValue()
                && quote3.getHigh() > CollectionUtilsJava.getFromEnd(screen_2_EMA13, 3).getValue();
        boolean secondCrossesEMA13 = quote2.getLow() < CollectionUtilsJava.getFromEnd(screen_2_EMA13, 2).getValue()
                && quote2.getHigh() > CollectionUtilsJava.getFromEnd(screen_2_EMA13, 2).getValue();
        boolean lastBelowEMA13 = quote1.getLow() < CollectionUtilsJava.getFromEnd(screen_2_EMA13, 1).getValue()
                && quote1.getHigh() < CollectionUtilsJava.getFromEnd(screen_2_EMA13, 1).getValue();
        if (thirdCrossesEMA13 && secondCrossesEMA13 && lastBelowEMA13) {
            LogJava.recordCode(LAST_BAR_BELOW_SCREEN_2, screen);
            LogJava.addDebugLine("The third and second crossed the EMA13, and the last is completely below");
            return new BlockResultJava(screen.getLastQuote(), LAST_BAR_BELOW_SCREEN_2);
        }
        return new BlockResultJava(screen.getLastQuote(), OK);
    }
}

File: src/main/kotlin/com/kishlaly/ta/analyze/tasks/blocks/two/ema/Short_ScreenTwo_EMA_LastBarTooLow.kt
package com.kishlaly.ta.analyze.tasks.blocks.two.ema

import com.kishlaly.ta.analyze.tasks.blocks.two.ScreenTwoBlock
import com.kishlaly.ta.model.*
import com.kishlaly.ta.model.indicators.EMA
import com.kishlaly.ta.utils.Log
import com.kishlaly.ta.utils.SymbolDataUtils

/**
 * to filter the situation when the third and second cross EMA13, and the last one is entirely lower (that is, the moment is already lost)
 */
class Short_ScreenTwo_EMA_LastBarTooLow : ScreenTwoBlock {
    override fun check(screen: SymbolData): BlockResult {
        val ema = SymbolDataUtils(screen, EMA::class.java)

        if (screen.lastQuote(3) crosses ema[3]
            && screen.lastQuote(2) crosses ema[2]
            && screen.lastQuote(1) below ema[1]
        ) {
            Log.recordCode(BlockResultCode.LAST_BAR_BELOW_SCREEN_2, screen)
            Log.addDebugLine("The third and second crossed the EMA13, and the last is completely below")
            return BlockResult(screen.lastQuote, BlockResultCode.LAST_BAR_BELOW_SCREEN_2)
        }

        return BlockResult(screen.lastQuote, BlockResultCode.OK)
    }
}